### DISCUSSION TITLE
tezedge + librustzcash (Sapling)

### DISCUSSION THREAD

**Discussion by storojs72** [tezedge/tezedge#792] [General]
Hi, folks! 

I'm spending a lot of time studying [ZCash](https://z.cash/technology/) elaborated technology based on [zk-SNARKs](https://z.cash/technology/zksnarks/) and it is awesome that Tezos [welcomes](https://decrypt.co/49912/tezos-will-get-zcashs-privacy-features-in-next-upgrade) adoption of Sapling protocol in order to improve its transactions privacy. I don't know the main business goals of tezedge project, but from engineering perspective it seems very interesting to integrate [librustzcash](https://github.com/zcash/librustzcash) directly into tezedge without having intermediate FFI API layer as it is in source code of [zcashd](https://github.com/zcash/zcash) node, since it is written in C++. As far as I know, main Tezos node is [implemented](https://gitlab.com/tezos/tezos/) in OCaml and it also has no escape from using external FFI API for OCaml <-> Rust communication. Obviously, tezedge being implemented on Rust doesn't require this stuff while integration with librustzcash's abstractions which simplifies versioning, compatibility and maintainability of the software component. 

Do You have mentioned tezedge + librustzcash (Sapling) integration in mind or in Your high-level design/business goals? Or even may be some work is already started in this direction (though I didn't find any branch in the repository with expressive name)? If so, I would probably like to contribute somehow to this activity if it is relevant.

**Comment by tizoc** [comment]
Hi @storojs72. The reason you couldn't find such code in the repository is that TezEdge doesn't directly use `librustzcash`, but only through the OCaml code that implements the Tezos economic protocol and it's environment (TezEdge is an implementation of the shell, but it shares the protocol code with Octez, with the FFI between Rust<->OCaml being handled by [ocaml-interop](https://github.com/tezedge/ocaml-interop)).

At the moment there are no plans for integrating `librustzcash` directly into TezEdge, because its use would still have to happen by calling OCaml code (the economic protocol code that calls the OCaml code implementing the bindings in [lib_sapling](https://gitlab.com/tezos/tezos/-/tree/master/src/lib_sapling)).

**Reply by storojs72** [reply]
@tizoc thanks for responding. So, You have actually Rust <-> Ocaml  <-> Rust complicated bindings in order to access `librustzcash` from `tezedge`, right? I suppose such bindings are not cheap - but probably its indeed a workaround if You don't want to port parts of mentioned economic part of Tezos from Ocaml to Rust

**Reply by tizoc** [reply]
That is correct. The important detail is that code in TezEdge is not a direct user of ZCash. If at some point we need to interact with ZCash directly from TezEdge we could (as you correctly pointed out) just avoid the intermediary OCaml code, but right now such interaction doesn't exist.
